### PR TITLE
recyclarr: init at 5.1.0

### LINKED ISSUE #239851: Package request: recyclarr
**Project description**

A command-line application that will automatically synchronize recommended settings from the [TRaSH guides](https://trash-guides.info/) to your Sonarr/Radarr instances.

**Metadata**

* homepage URL: https://recyclarr.dev/
* source URL: https://github.com/recyclarr/recyclarr
* license: mit
* platforms: linux

### PR DESCRIPTION
**Pull request by josephst**
adds Recyclarr, a CLI app for syncing settings to Sonarr & Radarr (usenet downloaders).

Closes #239851 

The binary is a single-file dotnet application, which includes a hard-coded dependence on `libicucore.dylib` on MacOS (https://github.com/dotnet/runtime/blob/1ffb77d7c5b5b380dfd8146858a044fadf6f21d4/src/native/libs/System.Globalization.Native/pal_icushim.c#L167)

This file is not included in the `icu` package;
therefore, it is necessary to add `darwin.ICU` to the search path on MacOS (with DYLD_LIBRARY_PATH)

###### Description of changes

[Recyclarr homepage](https://github.com/recyclarr/recyclarr)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - [x] and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by josephst** [comment]
Fixed white-space error

File: pkgs/tools/video/recyclarr/default.nix [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  nativeBuildInputs = [ makeWrapper ]
+    ++ lib.optional stdenv.hostPlatform.isLinux autoPatchelfHook
+    ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames;
+  buildInputs = [ icu zlib ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp recyclarr $out/bin
```
Comment (Janik-Haag) [inline pkgs/tools/video/recyclarr/default.nix:63]: does it work if you symlink the binary like for example here: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/tasks/filesystems/btrfs.nix#L70

File: pkgs/tools/video/recyclarr/default.nix [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+in
+stdenv.mkDerivation rec {
+  pname = "recyclarr";
+  version = "5.1.0";
+
+  src = fetchurl {
+    url = "https://github.com/recyclarr/recyclarr/releases/download/v${version}/recyclarr-${os}-${arch}.tar.xz";
+    inherit hash;
+  };
```
Comment (Janik-Haag) [inline pkgs/tools/video/recyclarr/default.nix:50]: Please use the `fetchFromGitHub` wrapper method https://nixos.org/manual/nixpkgs/stable/#fetchfromgithub

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -7989,6 +7989,12 @@
     github = "jorsn";
     githubId = 4646725;
   };
+  josephst = {
+    name = "Joseph Stahl";
+    email = "hello@josephstahl.com";
+    github = "josephst";
+    githubId = 1269177;
+  };
```
Comment (Janik-Haag) [inline maintainers/maintainer-list.nix:7997]: Please put this into it's own commit called: `maintainers: add josephst`

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
Thanks for packaging :heart: 

Can you fix the few nitpicks so we are good to merge?

File: pkgs/tools/video/recyclarr/default.nix (hunk 2 above)
Comment (josephst) [reply pkgs/tools/video/recyclarr/default.nix:50]: Is there a way to use `fetchFromGitHub` to get an archive from the Releases page? It's my understanding that `fetchFromGitHub` pulls a copy of the source code associated w/ a particular tag. What I'm aiming for here is downloading the compiled binaries from the Releases page (generated by the recyclarr repo CI). 

Right now, I'm using `fetchurl` as compiling from source 1) adds a significant amount of CPU/time/storage demands to download entire dotnet SDK for multiple platforms, and 2) requires patching several portions of source code related to the project trying to parse its own git history to generate a version string (these checks must be patched out since `fetchFromGitHub` does not pull the entire project's history).

File: pkgs/tools/video/recyclarr/default.nix (hunk 1 above)
Comment (josephst) [reply pkgs/tools/video/recyclarr/default.nix:63]: Can't get symlinking to work here - it builds but the target of the symlink (ie `src`) is no longer accessible after the build finishes. Keeping the `cp` command here so that the binary is copied into `$out` and accessible after build finishes/ is cleaned up. 

The usage here is modeled after some of the [trivial examples on nixpkgs manual](https://github.com/NixOS/nixpkgs/blob/e7a84def4794fb783ec907440541697e7b9c7945/pkgs/development/web/bun/default.nix#L47) as well as more complicated packages such as [radarr](https://github.com/NixOS/nixpkgs/blob/da0a385a5e720b8dfb422e4f6a1c466b001e4089/pkgs/servers/radarr/default.nix) which is a similar program to recyclarr (ie a C# program distributed as a binary from a GitHub release)

**Comment by josephst** [comment]
Should be ready for re-review. Added self as a maintainer in separate commit. Other issues (using `ln` instead of `cp` and using `fetchFromGitHub`) addressed above and left unchanged.

File: pkgs/tools/video/recyclarr/default.nix (hunk 2 above)
Comment (Janik-Haag) [reply pkgs/tools/video/recyclarr/default.nix:50]: Usually I would say building from source is the prefered way even if it is more compute intense, that's why we have cache.nixos.org and the part with it trying to read stuff from it's git history get's patched in many packages but I'm to deep out of my comfort zone to make a clear cut decision so I would rather have some one else review this again. Hope that's fine with you?

File: pkgs/tools/video/recyclarr/default.nix (hunk 1 above)
Comment (Janik-Haag) [reply pkgs/tools/video/recyclarr/default.nix:63]: I guess copy is fine then.

File: pkgs/tools/video/recyclarr/update.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -I nixpkgs=./. -i bash -p curl jq common-updater-scripts gnused nix coreutils
+
+set -euo pipefail
+
+latestVersion="$(curl -s "https://api.github.com/repos/recyclarr/recyclarr/releases?per_page=1" | jq -r ".[0].tag_name" | sed 's/^v//')"
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/update.sh:6]: API calls should optionally take a GITHUB_TOKEN because they are easily rate limited.

File: pkgs/tools/video/recyclarr/default.nix [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    cp recyclarr $out/bin
+    chmod +x $out/bin/recyclarr
+
+    runHook postInstall
+  '';
+
+  postInstall = ''
+    wrapProgram $out/bin/recyclarr \
+        --prefix PATH : ${lib.makeBinPath [git]} \
+        --prefix ${libPath}
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:72]:
```suggestion
      --prefix PATH : ${lib.makeBinPath [ git ]} \
      --prefix ${libPath}
```
we use 2 spaces in nix files

File: pkgs/tools/video/recyclarr/default.nix [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    ++ lib.optional stdenv.hostPlatform.isLinux autoPatchelfHook
+    ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames;
+  buildInputs = [ icu zlib ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp recyclarr $out/bin
+    chmod +x $out/bin/recyclarr
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:64]:
```suggestion
    install -Dm755 recyclarr -t $out/bin
```

File: pkgs/tools/video/recyclarr/default.nix [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  src = fetchurl {
+    url = "https://github.com/recyclarr/recyclarr/releases/download/v${version}/recyclarr-${os}-${arch}.tar.xz";
+    inherit hash;
+  };
+
+  sourceRoot = ".";
+
+  nativeBuildInputs = [ makeWrapper ]
+    ++ lib.optional stdenv.hostPlatform.isLinux autoPatchelfHook
+    ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames;
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:56]:
```suggestion
    ++ lib.optional stdenv.isLinux autoPatchelfHook
    ++ lib.optional stdenv.isDarwin fixDarwinDylibNames;
```
this is the same

File: pkgs/tools/video/recyclarr/default.nix [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    }."${stdenv.hostPlatform.system}"
+      or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
+
+  hash =
+    {
+      x64-linux_hash = "sha256-aEcinTJlO++rTeyqGJea0TWtleH6fyooA8RhT0Qj24c=";
+      arm64-linux_hash = "sha256-9Gyk09zAGzahP8FCGxj037vaK8h++3M5R2Qqop99Gs4=";
+      x64-osx_hash = "sha256-c87eOZBz+RtbIi+dlXKKVMyPI8JqYDuiaL4xOkDRFn0=";
+      arm64-osx_hash = "sha256-zSHgLXRDB6UA7V0LFgLq9ChqB40IHIJJxRqAYyVFlB8=";
+    }."${arch}-${os}_hash";
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:35]:
```suggestion
  hash = {
    x86_64-linux = "sha256-aEcinTJlO++rTeyqGJea0TWtleH6fyooA8RhT0Qj24c=";
    aarch64-linux = "sha256-9Gyk09zAGzahP8FCGxj037vaK8h++3M5R2Qqop99Gs4=";
    x86_64-darwin = "sha256-c87eOZBz+RtbIi+dlXKKVMyPI8JqYDuiaL4xOkDRFn0=";
    aarch64-darwin = "sha256-zSHgLXRDB6UA7V0LFgLq9ChqB40IHIJJxRqAYyVFlB8=";
  }."${stdenv.hostPlatform.system}";
```

File: pkgs/tools/video/recyclarr/default.nix [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    then "osx"
+    else "linux";
+  arch =
+    {
+      x86_64-linux = "x64";
+      aarch64-linux = "arm64";
+      x86_64-darwin = "x64";
+      aarch64-darwin = "arm64";
+    }."${stdenv.hostPlatform.system}"
+      or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:27]:
```suggestion
  arch = {
    x86_64-linux = "x64";
    aarch64-linux = "arm64";
    x86_64-darwin = "x64";
    aarch64-darwin = "arm64";
  }."${stdenv.hostPlatform.system}"
    or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
```

File: pkgs/tools/video/recyclarr/default.nix [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+, makeWrapper
+, autoPatchelfHook
+, fixDarwinDylibNames
+, darwin
+, recyclarr
+, git
+, icu
+, testers
+, zlib
+,
```
Comment (SuperSandro2000) [inline pkgs/tools/video/recyclarr/default.nix:13]:
```suggestion
```

**Comment by josephst** [comment]
Thanks for the reviews! Everything should be fixed in the most recent commits.

Also, ran the update script through shellcheck and made some of the suggested changes.

File: pkgs/tools/video/recyclarr/default.nix (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/tools/video/recyclarr/default.nix:13]: you forgot this

**Comment by josephst** [comment]
Fixed formatting, got merge conflict cleared up.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 240119` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>recyclarr</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]

File: pkgs/tools/video/recyclarr/default.nix (hunk 7 above)
Comment (ambroisie) [reply pkgs/tools/video/recyclarr/default.nix]: Should this not be using the target platform?

Comment (josephst) [reply pkgs/tools/video/recyclarr/default.nix]: @ambrosie I don't believe that's necessary here. The platform is not cross-compiled; Mac builds are built on MacOS and NixOS builds are built on NixOS. This makes `hostPlatform` and `buildPlatform` the same. `targetPlatform` [is a niche case](https://nixos.org/manual/nixpkgs/stable/#ssec-cross-platform-parameters) that's only relevant for compilers. 

Even if the cases were different, I think `hostPlatform` would be right since that refers to the platform that the code is running on (and thus the platform that patching binaries is necessary for). But here's it doesn't matter as host & build platforms are the same and `stdenv.isLinux` and `stdenv.isDarwin` can be used.
